### PR TITLE
system/term: Improve resolution for setting a valid PAM_TTY

### LINKED ISSUE #1593: PAM_TTY is not properly computed and exposed in situations where it could
**Describe the bug**

PAM_TTY is not properly computed by sudo-rs when input redirection or IO streams are unset or closed.

**To Reproduce**

With this minimal reproducer (that simulates what sshuttle does):

```python3
#!/usr/bin/env python3

import shutil
import socket
import subprocess
import sys

def main() -> int:
    sudo = shutil.which("sudo") or "sudo"
    #sudo = "/home/user-sudo/sudo"
    sudo = "/usr/lib/cargo/bin/sudo"
    #sudo = "/usr/bin/sudo.ws"
    cmd = [sudo, "-p", "[local sudo] Password: ", "id"]

    s1, s2 = socket.socketpair()

    def setup() -> None:
        s2.close()

    proc = subprocess.Popen(cmd, stdin=subprocess.DEVNULL, stdout=s1, preexec_fn=setup)
    s1.close()

    output = s2.makefile("rb").read()
    if output:
        sys.stdout.buffer.write(output)

    rc = proc.wait()
    s2.close()
    return rc

if __name__ == "__main__":
    raise SystemExit(main())
```

Modify `/etc/pam.d/sudo` so that it contains:

      auth optional pam_exec.so stdout /usr/bin/env

**Expected behavior**

PAM_SERVICE=sudo
PAM_USER=user-sudo
PAM_TTY=/dev/pts/1
PAM_RUSER=user-sudo
PAM_TYPE=auth

**Actual behavior**

PAM_SERVICE=sudo
PAM_USER=user-sudo
PAM_RUSER=user-sudo
PAM_TYPE=auth

So PAM_TTY is missing, even though we are in a terminal. This may make PAM modules not to properly handle things in the right way, and in particular authd will not run the right client for the conversation.

See also https://github.com/canonical/authd/issues/901
Also related to #1260

### PR DESCRIPTION
**Pull request by 3v1n0**
In sudo-rs we're just trying to use stdin as `PAM_TTY`, but this may not be available if some input is redirected.

For example, using tools such as `sshuttle` with `sudo-rs`, `PAM_TTY`, is currently unset while it's `/dev/pts/3` with sudo.ws

Now, the first commit would be enough to handle this in a simple way, but I also implemented a logic (for linux only, although BSDs can be added too), to follow the same logic that sudo is using to pick the TTY in a follow-up commit. Not sure if we have to go that deep though.

A simpler reproducer is (bug visible when uising `--close-stdin` or `--stdin-devnull`):

```python
#!/usr/bin/env python3
"""Minimal reproducer for sshuttle-style sudo launch."""

import argparse
import os
import shutil
import socket
import subprocess
import sys


def enable_sshuttle_syslog_fd_mode() -> subprocess.Popen[bytes]:
    """Mimic cmdline.py + ssyslog.py fd setup used by sshuttle --syslog."""
    sink = subprocess.Popen(
        ["cat"],
        stdin=subprocess.PIPE,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    assert sink.stdin is not None
    # ssyslog.close_stdin()
    os.close(0)
    # ssyslog.stdout_to_syslog() + stderr_to_syslog()
    os.dup2(sink.stdin.fileno(), 1)
    os.dup2(sink.stdin.fileno(), 2)
    return sink


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--close-stdin",
        action="store_true",
        help="close stdin before launching sudo (mimics sshuttle --syslog path)",
    )
    parser.add_argument(
        "--stdin-devnull",
        action="store_true",
        help="launch sudo with stdin from /dev/null",
    )
    parser.add_argument(
        "--sshuttle-syslog-fds",
        action="store_true",
        help="mimic sshuttle --syslog fd wiring before spawning sudo",
    )
    args = parser.parse_args()

    if args.sshuttle_syslog_fds and (args.close_stdin or args.stdin_devnull):
        parser.error("--sshuttle-syslog-fds cannot be combined with stdin flags")
    if args.close_stdin and args.stdin_devnull:
        parser.error("--close-stdin and --stdin-devnull are mutually exclusive")

    sudo = shutil.which("sudo") or "sudo"
    argv = [sudo, "-p", "[local sudo] Password: ", "id"]

    s1, s2 = socket.socketpair()

    def setup() -> None:
        # Run in child process.
        s2.close()

    stdin_obj = None
    syslog_sink = None
    if args.sshuttle_syslog_fds:
        syslog_sink = enable_sshuttle_syslog_fd_mode()
    elif args.close_stdin:
        os.close(0)
    elif args.stdin_devnull:
        stdin_obj = open(os.devnull, "rb")

    # We can't rely on stderr being tty-writable in syslog-fd mode.
    debug_fd = open("/dev/tty", "w", buffering=1) if os.path.exists("/dev/tty") else sys.__stderr__
    print(f"Starting command: {argv!r}", file=debug_fd)
    print(
        f"parent isatty: stdin={os.isatty(0)} stdout={os.isatty(1)} stderr={os.isatty(2)}",
        file=debug_fd,
    )
    proc = subprocess.Popen(
        argv,
        stdin=stdin_obj,
        stdout=s1,
        preexec_fn=setup,
    )
    s1.close()
    if stdin_obj is not None:
        stdin_obj.close()

    pfile = s2.makefile("rwb")
    output = pfile.read()
    if output:
        sys.stdout.buffer.write(output)
        sys.stdout.buffer.flush()

    rc = proc.wait()
    pfile.close()
    s2.close()
    if syslog_sink is not None:
        if syslog_sink.stdin is not None:
            syslog_sink.stdin.close()
        syslog_sink.wait()
    if debug_fd is not sys.__stderr__:
        debug_fd.close()
    return rc


if __name__ == "__main__":
    raise SystemExit(main())
```

Closes: #1593

### CODE REVIEW

File: src/system/term/mod.rs [hunk 1]
```diff
@@ -254,7 +262,99 @@ impl<F: AsFd> Terminal for F {
...
+        std::io::stderr().ttyname(),
+    ]
+    .into_iter()
+    .flatten()
+    .next()
+    {
+        return Ok(tty_name);
+    }
+
+    Err(io::ErrorKind::Unsupported.into())
```
Comment (squell) [inline src/system/term/mod.rs]: Why not simply
```suggestion
    std::io::stdin().tty_name()
        .or_else(|_| std::io::stdout().tty_name())
        .or_else(|_| std::io::stderr().tty_name())
```

**Review by squell — COMMENTED** [review]
I would like to have some version of this PR in the next release.

I think I agree with the initial comment that the "simple solution" with the fix of the first commit (with associated integration tests) suffices (until we have better information why it wouldn't).

ttyname_from_dev_linux seems to me to try the following three things:

- First: stdin/stdout/stderr in order
- Then: /dev/console
- Finally: iterate over /dev/pts

In the Rust bits, we already check stdin/stdout/stderr (although we use different code), so that feels duplicate; I think checking /dev/console is maybe fine as well (for Linux), although the use cases are probably rare. Note that a few weeks ago, in ogsudo Todd moved the order around so `/dev/console` is checked **first** (that was after this PR was opened).

For the /dev/pts loop in `find_tty_in_dir`, it feels a bit to me that if sudo  needs to start guessing what the controlling TTY is, we might as well leave PAM_TTY unset and decide that it's PAM's job, so I would prefer to keep that out.

Also, simply resolving `/dev/tty` (or even starting with that) would feel logical here; but I would need to dig a bit deeper to know why we aren't doing that in the first place. (Maybe `ogsudo` can't do that because it has to service more platforms, but we only do FreeBSD and Linux and both of those have `/dev/tty`

**Comment by bjorn3** [comment]
> Also, simply resolving /dev/tty (or even starting with that) would feel logical here; but I would need to dig a bit deeper to know why we aren't doing that in the first place. (Maybe ogsudo can't do that because it has to service more platforms, but we only do FreeBSD and Linux and both of those have /dev/tty

It does check that the device minor id matches between the controlling tty and the `/dev/pts` entry to ensure they are the same tty.

**Comment by squell** [comment]
Also turns out that `/dev/tty` handily resolves to `/dev/tty` so no help there.

**Comment by 3v1n0** [comment]
> I think I agree with the initial comment that the "simple solution" with the fix of the first commit (with associated integration tests) suffices (until we have better information why it wouldn't).

As mentioned in the commit, the initial fix is fine to handle some basic cases, but other tests such as [`6383c09` (this PR)](https://github.com/trifectatechfoundation/sudo-rs/pull/1591/changes/6383c098d004a98c2bc07ec414916c240546902f) woudl not pass otherwise (while they do in og-sudo).

**Comment by squell** [comment]
Thanks for the comments!

Based on this discussion: I have some time this week to fiddle around with the changes this PR introduces, which I'll do then; that might turn into a commit or two on top of the existing ones (which I'll then ask Bjorn to review) or a PR targetting @3v1n0's branch, or simply a decision to approve this PR as-is. I think that's more constructive on my side and a more efficient use of @3v1n0's time than doing two or three rounds of back and forth in GitHub. :-)

**Comment by 3v1n0** [comment]
> Thanks for the comments!
> 
> Based on this discussion: I have some time this week to fiddle around with the changes this PR introduces, which I'll do then; that might turn into a commit or two on top of the existing ones (which I'll then ask Bjorn to review) or a PR targetting @3v1n0's branch, or simply a decision to approve this PR as-is. I think that's more constructive on my side and a more efficient use of @3v1n0's time than doing two or three rounds of back and forth in GitHub. :-)

As you like, feel free to refactor my commits yourself if you want if there are not big changes, rather than adding fixup-commits on top of the branch.

**Comment by squell** [comment]
I've kept the original logic more or less in place but rewritten it to be more high-level. Also (in the last commit) skips the "fallback" stdin/stdout/stderr" check now *if* a TTY does exist, but the code simply hasn't found it. It's probably better to not set PAM_TTY rather than set it to something incorrect.

**Review by squell — DISMISSED** [review]

File: src/system/term/find_tty.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+};
+
+use crate::system::{Process, WithProcess, term::Terminal};
+
+pub(super) fn ttyname_from_dev() -> io::Result<Option<OsString>> {
+    let Ok(Some(tty_dev)) = Process::tty_device_id(WithProcess::Current) else {
+        return Ok(None);
+    };
+
+    let tty_dev = tty_dev.inner();
```
Comment (bjorn3) [inline src/system/term/find_tty.rs]: Maybe pass `DeviceId` around instead?

File: src/system/term/find_tty.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    if is_our_tty(metadata, tty_dev) {
+        Some(path.into())
+    } else {
+        None
+    }
+}
+
+fn find_tty_in_dir(dir: &Path, tty_dev: libc::dev_t) -> Option<OsString> {
+    for entry in fs::read_dir(dir).ok()?.flatten() {
```
Comment (bjorn3) [inline src/system/term/find_tty.rs]: Took me looking at the docs before I managed to wrap my head around the `.flatten()` here. Turns out `Result` implements `IntoIterator` by returning an iterator which yields the Ok value or nothing in case of an error. Maybe `.filter_map(|entry| entry.ok())` is clearer?

File: src/system/term/find_tty.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+fn dev_check(path: &Path, tty_dev: libc::dev_t) -> Option<OsString> {
+    let metadata = fs::metadata(path).ok()?;
+
+    if is_our_tty(metadata, tty_dev) {
+        Some(path.into())
+    } else {
+        None
+    }
```
Comment (bjorn3) [inline src/system/term/find_tty.rs]:
```suggestion
    is_our_tty(metadata, tty_dev).then(|| path.into())
```

is a bit shorter.

File: src/system/term/find_tty.rs (hunk 2 above)
Comment (squell) [reply src/system/term/find_tty.rs]: I considered that, but then we get a lot of `.inner()` and the functions taking the `dev_t` are not `pub` so I decided against that.

File: src/system/term/find_tty.rs (hunk 3 above)
Comment (squell) [reply src/system/term/find_tty.rs]: Exactly.

But if *you* need to wrap your head around this, then clearly it is clearer to do use `filter_map`.

File: src/system/term/find_tty.rs (hunk 2 above)
Comment (squell) [reply src/system/term/find_tty.rs]: Of course after all the refactors it can be done since it just moves one `.inner()` to another spot.

**Comment by 3v1n0** [comment]
I've added further tests to cover other edge cases.

**Comment by squell** [comment]
@3v1n0 By the way, wouldn’t the present solution also work well for modern FreeBSD? I.e. is the Linux cfg necessary? I haven’t looked at this situation yet but maybe you have.

**Comment by 3v1n0** [comment]
> @3v1n0 By the way, wouldn’t the present solution also work well for modern FreeBSD? I.e. is the Linux cfg necessary? I haven’t looked at this situation yet but maybe you have.

I did not run it there, and given that indeed the exec module has some differences I had no time to try running it there, then I preferred to be conservative.

We can still enable them and see if someone with more bsd experience wants to re-iterate.

**Comment by squell** [comment]
Overall logic (check /dev/console, check the std streams, look into /dev/pts) looks to me like it should be identical. I'll check the `pam_tty` tests on FreeBSD.

**Comment by squell** [comment]
I've manually tested the changes on FreeBSD and they work, so the cfg(target_os = linux) can be safely removed in the production code.

However, in the integration tests, `pam_exec` on FreeBSD doesn't have the "log=" parameter but I think the test can probably be achieved by examing the environment from within an executed script.

I can try to alter the tests so they work on FreeBSD. I might fail due to time constraints; in that case I recommend we remove the cfg(target_os = linux) from the production code anyway.

**Comment by squell** [comment]
That wasn't too hard. Of course to run the tests on FreeBSD requires a rebase on current main. We don't have a CI option for the compliance tests on FreeBSD, but I vouch for the fact that with 1bdc10a the `pam_tty` tests all succeeded with both `SUDO_UNDER_TEST=ours` and `theirs`.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs [hunk 5]
```diff
@@ -1,11 +1,87 @@
...
+use sudo_test::{Command, Directory, Env, User};
 
 use crate::{PASSWORD, USERNAME};
 
 mod env;
 
+const TEST_ENV_EXPECTED_TTY: &str = "SUDO_RS_TEST_ENV_EXPECTED_TTY";
+
+fn test_temp_paths(test_name: &str) -> (String, String, String) {
+    let base = std::env::temp_dir().join(format!("sudo-rs-pam-{test_name}-{}", std::process::id()));
```
Comment (bjorn3) [inline test-framework/sudo-compliance-tests/src/sudo/pam.rs]: This doesn't need to be unique. All tests run inside unique docker containers anyway.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs [hunk 6]
```diff
@@ -102,3 +178,301 @@ fn sudo_dash_i_uses_correct_service_file() {
...
+printf '%s=%s\n' '{TEST_ENV_EXPECTED_TTY}' "$expected_tty" >&3
+cat {pam_env_value} >&3"#
+        ))
+        .as_user(USERNAME)
+        .tty(true)
+        .output(&env)
+        .stdout();
+
+    let (expected, pam_env) = parse_expected_tty_and_pam_env(&stdout);
+    assert_pam_tty_matches_expected(&expected, &pam_env);
```
Comment (bjorn3) [inline test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: Maybe these should be merged into a single function?

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs [hunk 7]
```diff
@@ -102,3 +178,301 @@ fn sudo_dash_i_uses_correct_service_file() {
...
+    let (tmp_dir, pam_env_value, repro_log) = test_temp_paths("stdio-fds-not-ttys");
+
+    let env = build_pam_capture_env(tmp_dir.as_str(), pam_env_value.as_str());
+
+    let stdout = Command::new("sh")
+        .arg("-c")
+        .arg(format!(
+            r#"exec 3>&1
+expected_tty=$(tty <&3)
+exec </dev/null >{repro_log} 2>&1
```
Comment (bjorn3) [inline test-framework/sudo-compliance-tests/src/sudo/pam.rs]:
```suggestion
exec </dev/null >/dev/null 2>&1
```

this file is deleted after the test has finished anyway.

**Comment by 3v1n0** [comment]
> That wasn't too hard. Of course to run the tests on FreeBSD requires a rebase on current main. We don't have a CI option for the compliance tests on FreeBSD, but I vouch for the fact that with [1bdc10a](https://github.com/trifectatechfoundation/sudo-rs/commit/1bdc10a7c74c3ed8b17cd72b78fc8e871912ed8f) the `pam_tty` tests all succeeded with both `SUDO_UNDER_TEST=ours` and `theirs`.

Ok I can do the rebase if it's not an issue.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 5 above)
Comment (3v1n0) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs]: Yeah, but all the test run in the same container, no?

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 6 above)
Comment (3v1n0) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: I preferred to keep it separated so that we have two different phases, and if in future we need to check other PAM environment variables it's already there.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 5 above)
Comment (bjorn3) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs]: They run in different containers.

I can fix my review comments and have Marc review them if you want. The non-test half of this PR LGTM already.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 6 above)
Comment (bjorn3) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: How about doing

```rust
let pam_env = parse_pam_env(stdout);
assert_pam_tty_matches_expected(&pam_env);
```

and moving the `SUDO_RS_TEST_ENV_EXPECTED_TTY` reading into `assert_pam_tty_matches_expected`? Currently the expected and actual tty reading are split into two functions.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 7 above)
Comment (3v1n0) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs]: done

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 6 above)
Comment (3v1n0) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: Yes, but then it makes it harder to do `pam_tty.starts_with("/dev/pts/")` or similar checks.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 5 above)
Comment (bjorn3) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs]: I've pushed a commit to remove `test_temp_paths` and instead hard code the env capture path.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 6 above)
Comment (bjorn3) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: True. Let's leave it as is then.

**Review by bjorn3 — APPROVED** [review]
Maybe putting all the PAM_TTY tests in a separate file would be a bit nicer, but not worth blocking merging this PR on.

File: test-framework/sudo-compliance-tests/src/sudo/pam.rs (hunk 6 above)
Comment (3v1n0) [reply test-framework/sudo-compliance-tests/src/sudo/pam.rs:199]: Ok droppped that though so now we could move to something like that.

**Comment by squell** [comment]
After the rebase/force push I believe the cfg(target_os = "linux") has reappeared, also the compliance tests are failing again on FreeBSD.

(trying to work out what's wrong)

Cherry-picked https://github.com/trifectatechfoundation/sudo-rs/commit/1bdc10a7c74c3ed8b17cd72b78fc8e871912ed8f on top of this one and that works.

**Review by squell — APPROVED** [review]

**Review by bjorn3 — APPROVED** [review]

**Comment by squell** [comment]
Thanks @3v1n0 and @bjorn3 for working together to pull this over the finish line. A round of (self-)congratulation for the collaboration is in order here.

Ottimo lavoro, grazie per la pazienza.
